core/store/src/trie/prefetching_trie_storage.rs

```diff
@@ -0,0 +1,399 @@
+/// independent of the prefetcher.
+#[derive(Default, Clone)]
+pub(crate) struct PrefetchStagingArea(Arc<Mutex<InnerPrefetchStagingArea>>);
+
+#[derive(Default)]
+struct InnerPrefetchStagingArea {
+    slots: HashMap<CryptoHash, PrefetchSlot>,
+    size_bytes: usize,
+}
+
+/// Result when atomically accessing the prefetch staging area.
+pub(crate) enum PrefetcherResult {
+    SlotReserved,
+    Pending,
+    Prefetched(Arc<[u8]>),
+    MemoryLimitReached,
+}
+
+/// Type used interanlly in the staging area to keep track of requests.
+#[derive(Clone, Debug)]
+enum PrefetchSlot {
+    PendingPrefetch,
+    PendingFetch,
+    Done(Arc<[u8]>),
+}
+
+impl TrieStorage for TriePrefetchingStorage {
+    // Note: This is the tricky bit of the implementation.
+    // We have to retrieve data only once in many threads, so all IO threads
+    // have to go though the staging area and check for inflight requests.
+    // The shard cache mutex plus the prefetch staging area mutex are used for
+    // that in combination. Let's call the first lock S and the second P.
+    // The rules for S and P are:
+    // 1. To avoid deadlocks, S must always be requested before P, if they are
+    //    held at the same time.
+    // 2. When looking up if something is already in the shard cache, S must not
+    //    be released until the staging area is updated by the current thread.
+    //    Otherwise, there will be race conditions that could lead to multiple
+    //    threads looking up the same value from DB.
+    // 3. IO threads should release S and P as soon as possible, as they can
+    //    block the main thread otherwise.
+    fn retrieve_raw_bytes(&self, hash: &CryptoHash) -> Result<Arc<[u8]>, StorageError> {
+        // Try to get value from shard cache containing most recently touched nodes.
+        let mut shard_cache_guard = self.shard_cache.0.lock().expect(POISONED_LOCK_ERR);
+        if let Some(val) = shard_cache_guard.get(hash) {
+            return Ok(val);
+        }
+
+        // If data is already being prefetched, wait for that instead of sending a new request.
+        let prefetch_state = PrefetchStagingArea::get_and_set_if_empty(
+            &self.prefetching,
+            hash.clone(),
+            PrefetchSlot::PendingPrefetch,
+        );
+        // Keep lock until here to avoid race condition between shard cache insertion and reserving prefetch slot.
+        std::mem::drop(shard_cache_guard);
+
+        match prefetch_state {
+            PrefetcherResult::SlotReserved => {
+                let key = TrieCachingStorage::get_key_from_shard_uid_and_hash(self.shard_uid, hash);
+                let value: Arc<[u8]> = self
+                    .store
+                    .get(DBCol::State, key.as_ref())
+                    .map_err(|_| StorageError::StorageInternalError)?
+                    .ok_or_else(|| {
+                        StorageError::StorageInconsistentState("Trie node missing".to_string())
+                    })?
```

Comment:
You are right, this isn't ideal. Error are unrecoverable at this point but we don't want to end up in a infinite loop anyway.
The fix I just pushed handles `None` and `Err` separately and has some comments. And we release the slot such that any thread waiting on the value can progress. Additionally, the main thread will try again fetching the data on its own, just in case something is fishy with the store in the prefetcher.